### PR TITLE
Record latency of imap commands in imap nio client.

### PR DESCRIPTION
**Pull request by davis-thomas**
Record latency of imap commands in imap nio client.

## Description
Need to know the latency of each imap command executed by the imap-nio client. Changes include recording start end end time of an imap command and storing the latency in ImapAsyncResponse.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Major release (change is NOT backward compatible with prior release)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java [hunk 1]
```diff
@@ -261,6 +266,13 @@ public void recordResponseBytes(final int length) {
             responseTotalBytes += length;
         }
 
+        /**
+         * @return request start time.
+         */
+        public long getRequestStartTimeInMillis() {
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:272]: what is the reason of not using requestSentTime which is already there?

File: core/src/test/java/com/yahoo/imapnio/async/client/ImapFutureTest.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public void beforeMethod() throws IOException, ProtocolException, ImapAsyncClien
         final ImapRequest imapRequest = new AuthXoauth2Command("username", "token", new Capability(capas));
         final int requestTotalBytes = "a1".getBytes(StandardCharsets.US_ASCII).length + 1 + imapRequest.getCommandLineBytes().readableBytes();
         imapAsyncResp = new ImapAsyncResponse(imapRequest.getCommandType(), requestTotalBytes,
-                oneImapResponse.toString().getBytes(StandardCharsets.US_ASCII).length + 2, imapResponses);
+                oneImapResponse.toString().getBytes(StandardCharsets.US_ASCII).length + 2, imapResponses, null);
```
Comment (nhant01) [inline core/src/test/java/com/yahoo/imapnio/async/client/ImapFutureTest.java:63]: Create the total elapsed time for the IMAPResponse and perform an assertion to validate the elapsed time.

File: core/src/test/java/com/yahoo/imapnio/async/response/ImapAsyncResponseTest.java [hunk 3]
```diff
@@ -34,7 +34,7 @@ public void testImapAsyncResponse() throws IOException, ProtocolException, ImapA
         final IMAPResponse oneImapResponse = new IMAPResponse("a1 OK CAPABILITY completed");
         imapResponses.add(oneImapResponse);
         final int responseTotalBytes = oneImapResponse.toString().getBytes(StandardCharsets.US_ASCII).length + 2;
-        final ImapAsyncResponse resp = new ImapAsyncResponse(imapRequest.getCommandType(), requestTotalBytes, responseTotalBytes, imapResponses);
+        final ImapAsyncResponse resp = new ImapAsyncResponse(imapRequest.getCommandType(), requestTotalBytes, responseTotalBytes, imapResponses, null);
```
Comment (nhant01) [inline core/src/test/java/com/yahoo/imapnio/async/response/ImapAsyncResponseTest.java:37]: Same here.

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java [hunk 4]
```diff
@@ -560,8 +573,9 @@ public <T> void handleChannelResponse(@Nonnull final IMAPResponse serverResponse
                     }
                 }
                 // see rfc3501, page 63 for details, since we always give a tagged command, response completion should be the first tagged response
+                final long totalTimeElapsedInMillis = System.currentTimeMillis() - curEntry.getRequestStartTimeInMillis();
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:576]: final long totalTimeElapsedInMillis = clock.millis() - curEntry.getRequestSentTime();

File: core/src/main/java/com/yahoo/imapnio/async/response/ImapAsyncResponse.java [hunk 5]
```diff
@@ -26,13 +28,15 @@ public class ImapAsyncResponse {
      * @param requestTotalBytes number of bytes in request
      * @param responseTotalBytes number of bytes in response
      * @param responses list of response lines
+     * @param totalTimeElapsedInMillis total time elapsed in millis
      */
     public ImapAsyncResponse(final ImapCommandType commandType, final int requestTotalBytes, final int responseTotalBytes,
-            final Collection<IMAPResponse> responses) {
+            final Collection<IMAPResponse> responses, long totalTimeElapsedInMillis) {
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/response/ImapAsyncResponse.java:34]: final long totalTimeElapsedInMillis

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java (hunk 1 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:272]: Fan, is this the time recorded when the command submitting to the server?

Comment (davis-thomas) [reply core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:272]: Looks like requestSentTime doesn't really point to the time time when request is being sent. Had checked with Vikram as well.

**Review by fansu — CHANGES_REQUESTED** [review]
Please also upgrade patch version (the 3rd number in X.X.X) in pom.xml, core/pom.xml and README

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java (hunk 1 above)
Comment (fansu) [reply core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:272]: I see it is fine then. keep the requestStartTimeInMillis, I will delete requestSentTime.

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java [hunk 6]
```diff
@@ -335,7 +347,7 @@ public ImapFuture<ImapAsyncResponse> execute(@Nonnull final ImapRequest command)
         final ImapFuture<ImapAsyncResponse> cmdFuture = new ImapFuture<ImapAsyncResponse>();
         final String tag = getNextTag();
         final int requestTotalBytes = tag.getBytes(StandardCharsets.US_ASCII).length + SPACE_LENGTH + command.getCommandLineBytes().readableBytes();
-        requestsQueue.add(new ImapCommandEntry(command, cmdFuture, tag, requestTotalBytes));
+        requestsQueue.add(new ImapCommandEntry(command, cmdFuture, tag, requestTotalBytes, System.currentTimeMillis()));
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:350]: clock.millis()

Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:350]: clock.millis()

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java (hunk 1 above)
Comment (fansu) [reply core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:272]: Change line 241 to append(getRequestStartTimeInMillis()) as well.

**Review by nhant01 — COMMENTED** [review]
README needs an update to use the 5.0.14 version.
<version>5.0.14</version>

File: pom.xml [hunk 7]
```diff
@@ -5,7 +5,7 @@
     <groupId>com.yahoo.imapnio</groupId>
     <artifactId>imapnio</artifactId>
     <packaging>pom</packaging>
-    <version>5.0.13</version>
+    <version>5.0.14</version>
```
Comment (fansu) [inline pom.xml:8]: please fix README as well

**Review by fansu — COMMENTED** [review]
Overall good, Fix readme and also resolve conflict.

File: core/src/main/java/com/yahoo/imapnio/async/response/ImapAsyncResponse.java [hunk 8]
```diff
@@ -18,6 +18,8 @@ public class ImapAsyncResponse {
     private int responseTotalBytes;
     /** List of IMAPResponse lines. */
     private Collection<IMAPResponse> responses;
+    /** Total time elapsed by command in millis. */
+    private long totalTimeElapsedInMillis;
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/response/ImapAsyncResponse.java:22]: commandExecutionTime.

**Review by nvsnvikram — APPROVED** [review]

**Review by fansu — APPROVED** [review]

**Pull request merged**